File: packages/ios-sdk/Sources/PaaqIntelligence/PaaqTouchTracker.swift
import Foundation
#if canImport(UIKit)
import UIKit

private let rageTapWindowSec: TimeInterval = 0.8
private let rageTapMinCount = 3
private let rageTapRadiusPt: CGFloat = 40
private let deadTapDelaySec: TimeInterval = 2.5

/// Global tap observer, installed once per process via method swizzling on
/// `UIApplication.sendEvent(_:)` — the same technique most RUM/analytics SDKs
/// use for automatic touch capture, since iOS has no per-view-agnostic touch
/// hook the way Android's Window.Callback or Flutter's GestureBinding do.
/// Only observes; the original sendEvent implementation always still runs.
final class PaaqTouchTracker {
    static let shared = PaaqTouchTracker()

    var onRageTap: ((CGPoint, Int) -> Void)?
    var onDeadTap: ((CGPoint) -> Void)?
    var lastSignalAt: () -> Date = { .distantPast }

    private var recentTaps: [(time: Date, point: CGPoint)] = []
    private var rageCooldownUntil = Date.distantPast
    private var installed = false

    func install() {
        guard !installed else { return }
        installed = true
        let cls = UIApplication.self
        let originalSelector = #selector(UIApplication.sendEvent(_:))
        let swizzledSelector = #selector(UIApplication.paaq_sendEvent(_:))
        guard
            let originalMethod = class_getInstanceMethod(cls, originalSelector),
            let swizzledMethod = class_getInstanceMethod(cls, swizzledSelector)
        else { return }
        method_exchangeImplementations(originalMethod, swizzledMethod)
    }

    fileprivate func handle(touches: Set<UITouch>) {
        guard let touch = touches.first, touch.phase == .began else { return }
        let point = touch.location(in: touch.window)
        let now = Date()

        recentTaps.removeAll { now.timeIntervalSince($0.time) > rageTapWindowSec }
        recentTaps.append((now, point))
        let cluster = recentTaps.filter { hypot($0.point.x - point.x, $0.point.y - point.y) <= rageTapRadiusPt }
        if cluster.count >= rageTapMinCount && now > rageCooldownUntil {
            rageCooldownUntil = now.addingTimeInterval(rageTapWindowSec)
            recentTaps.removeAll()
            onRageTap?(point, cluster.count)
        }

        // No cheap way to hit-test "was this a real control" on iOS without
        // walking the view hierarchy, so dead-tap detection here is purely
        // temporal (no signal followed the tap) — a looser heuristic than
        // the web SDK's, documented as such.
        let tapAt = now
        DispatchQueue.main.asyncAfter(deadline: .now() + deadTapDelaySec) { [weak self] in
            guard let self else { return }
            if self.lastSignalAt() < tapAt {
                self.onDeadTap?(point)
            }
        }
    }
}

private extension UIApplication {
    @objc func paaq_sendEvent(_ event: UIEvent) {
        if event.type == .touches, let touches = event.allTouches {
            PaaqTouchTracker.shared.handle(touches: touches)
        }
        // Selectors were swapped by `install()`, so this calls the real
        // original `sendEvent(_:)` implementation.
        self.paaq_sendEvent(event)
    }
}
#endif

File: packages/android-sdk/src/main/kotlin/io/paaq/intelligence/PaaqTouchTracker.kt
package io.paaq.intelligence

import android.os.Handler
import android.os.Looper
import android.os.SystemClock
import android.view.MotionEvent
import android.view.Window
import kotlin.math.hypot

private const val RAGE_TAP_WINDOW_MS = 800L
private const val RAGE_TAP_MIN_COUNT = 3
private const val RAGE_TAP_RADIUS_PX = 120f // generous, since px-per-dp varies across devices
private const val DEAD_TAP_DELAY_MS = 2500L

/**
 * Global tap observer for one Activity — Window.Callback is the same hook
 * Android itself uses to dispatch touches to views, so wrapping it lets us
 * see every tap automatically, with no per-view opt-in required (unlike
 * scroll/form tracking, which have no equivalent universal hook and stay
 * opt-in). We only observe; dispatchTouchEvent's return value always comes
 * from the real callback so app behavior is untouched.
 *
 * Unlike the web SDK, this has no cheap way to hit-test "was this a real
 * button" — so dead-tap detection here is purely temporal (no signal
 * followed the tap), a looser heuristic than web's, documented as such.
 */
internal class PaaqTouchTracker(
    private val onRageTap: (x: Float, y: Float, count: Int) -> Unit,
    private val onDeadTap: (x: Float, y: Float) -> Unit,
    private val lastSignalAtMs: () -> Long,
) {
    private data class Tap(val time: Long, val x: Float, val y: Float)
    private val recentTaps = mutableListOf<Tap>()
    private var rageCooldownUntil = 0L
    private val handler = Handler(Looper.getMainLooper())

    fun onTouchEvent(event: MotionEvent) {
        if (event.action != MotionEvent.ACTION_DOWN) return
        val now = SystemClock.elapsedRealtime()
        val x = event.rawX
        val y = event.rawY

        recentTaps.removeAll { now - it.time > RAGE_TAP_WINDOW_MS }
        recentTaps.add(Tap(now, x, y))
        val cluster = recentTaps.filter { hypot((it.x - x).toDouble(), (it.y - y).toDouble()) <= RAGE_TAP_RADIUS_PX }
        if (cluster.size >= RAGE_TAP_MIN_COUNT && now > rageCooldownUntil) {
            rageCooldownUntil = now + RAGE_TAP_WINDOW_MS
            recentTaps.clear()
            onRageTap(x, y, cluster.size)
        }

        val tapAtMs = System.currentTimeMillis()
        handler.postDelayed({
            if (lastSignalAtMs() < tapAtMs) onDeadTap(x, y)
        }, DEAD_TAP_DELAY_MS)
    }
}

internal class PaaqWindowCallbackWrapper(
    private val delegate: Window.Callback,
    private val tracker: PaaqTouchTracker,
) : Window.Callback by delegate {
    override fun dispatchTouchEvent(event: MotionEvent): Boolean {
        tracker.onTouchEvent(event)
        return delegate.dispatchTouchEvent(event)
    }
}
